### PR TITLE
Get server concurrency issue

### PR DESCRIPTION
**Pull request by alkiskal**
Fix for [#311](https://github.com/searchbox-io/Jest/issues/311).
Includes a simple implementation for an immutable circular list.
Although just synchronising the access method would work too, it would produce additional overhead.

### CODE REVIEW

File: jest-common/src/main/java/io/searchbox/client/AbstractJestClient.java [hunk 1]
```diff
@@ -115,13 +115,53 @@ public ServerPool(final Set<String> servers) {
...
+                current.next = nextNode;
+                current = nextNode;
+            }
+            current.next = head;
+            current = head;
+        }
+
+        public E next() {
+            if(current == null) return null;
+            E data = current.item;
```
Comment (jasonculverhouse) [inline jest-common/src/main/java/io/searchbox/client/AbstractJestClient.java]: Do you care if 2 successive calls get the same server or a server in the ring is skipped?

    
    final ImmutableList<String> serversRing;
    final AtomicInteger current = new AtomicInteger();
    
    public ServerPool(final Set<String> servers) {
        ...
        serversRing = new ImmutableList.copyOf(servers);
        ...
    }
    
    public String getNextServer() {
        Preconditions.checkState(!servers.isEmpty(), "Server list is empty");
        return serversRing.get(current.getAndIncrement() % serversRing.size());
    }

Comment (alkiskal) [inline jest-common/src/main/java/io/searchbox/client/AbstractJestClient.java]: This is a slick solution. Doesn't it add a performance hit though? As for the skipping or getting the same server for consecutive calls, I don't know the internals of Elastic but would it cause a performance hit? By synchronising the `getNextServer` method then these problems would vanish too. Moreover, depending on the use case sync could outperform atomic or vice versa according to this http://stackoverflow.com/questions/3556283/in-java-what-is-the-performance-of-atomicinteger-compareandset-versus-synchron. What are your thoughts?

Comment (jasonculverhouse) [inline jest-common/src/main/java/io/searchbox/client/AbstractJestClient.java]: I think this is as fast as you are going to get without a side effect https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/atomic/package-summary.html

You must have some performance measurement perhaps you could share

BTW, This is the "implementation" of `Iterators.cycle()` that is causing the issue.  It has the same issue as this code, it just manifests itself in a different way. The iterator is exhausted at `!iterator.hasNext())` but one thread is calling `iterator.next()` before the assignment of `iterator = iterable.iterator();` which restarts the cycle.

      @Override
      public boolean hasNext() {
        if (!iterator.hasNext()) {
          iterator = iterable.iterator();
        }
        return iterator.hasNext();
      }

Comment (alkiskal) [inline jest-common/src/main/java/io/searchbox/client/AbstractJestClient.java]: Yes I agree, the solution you provided is much much better. So what goes? I mean would you like me to change the implementation or would you prefer to make another pull request (I would close mine). I don't mind either way.

**Comment by joeslice** [comment]
@alkiskal I have created a commit to the effect from above:
https://github.com/joeslice/Jest/commit/0f93e891831169b745fbabe7bff9a2fdaee462be

Please feel free to bring it into this commit.  If you would prefer, I can submit it in a separate PR, or a PR to your PR :)

**Comment by jasonculverhouse** [comment]
Hey, sorry looking at this in the phone but @joeslice your code has
concurrency issues too that are much worse since it will take many I
wearing to recover.

You can read the value from the atomic integer into a local variable and
the add the negative value or else you can potientially add -(n) and -(n+1)
and then your index will be -(2n+1).  If you read into a local variable you
need to use compareAndSet

You can look at my solution the worst that will happen is integer overflow
which will wrap.

Jason


Jason

**Comment by joeslice** [comment]
@jasonculverhouse Thanks for the feedback.  I was modeling my commit after your comment above, but merely wanted to add the overflow protection (and a test etc).  

I'm not immediately seeing the concern you are raising about -(2n+1).  Would you mind taking another quick look at the commit?  I'm pretty sure the section below is what you were describing in your comment...?

```java
if (nextIndex == resetAt) {
    currentIndex.addAndGet(-resetAt);
}
```

We could do `currentIndex.addAndGet(-nextIndex)`, but that's pretty much the same thing.

**Comment by jasonculverhouse** [comment]
Pretend 10 servers...  ( don't think (-2(n+1) can happen looking at this but...) 

    if (serverList.size() > 0) {
        final int nextIndex = currentIndex.getAndIncrement();

At this point n threads can have incremental values "interleaved overflow" where you code never expects this to be `> resetAt`
thread one nextIndex == 10
thread two nextIndex == 11

        if (nextIndex == resetAt) {

only one resets the value

            currentIndex.addAndGet(-resetAt);
        }

if this happens then this `%` causes servers repeated servers to be selected since the `nextIndex` isn't reliably reset to `0`  `-resetAt` because you can have an interleaved overflow

        return serverList.get(nextIndex % serverList.size());
    }
    else throw new NoServerConfiguredException("No Server is assigned to client to connect");


With the single statement the ring selection is always sequential even under the overflow condition because MAX_INT becomes MIN_INT

    return serversRing.get(current.getAndIncrement() % serversRing.size());

**Comment by joeslice** [comment]
@jasonculverhouse I appreciate you getting back to me so quickly.  One thing I was trying to ensure is that we never use a negative index to `serversRing.get()` or `serverList.get()`.  A bit of paranoia as an overflow is not going to happen for a long time (over a year even at 50 queries per second!) so maybe not even necessary (although crazier things have happened).  I arbitrarily selected 1MM, but obviously it doesn't much matter what that value is.

As to the interleaved overflow, my intent indeed is that a single iteration will subtract the `resetAt` value (hence the `==` check and the `addAndGet(-resetAt)`, leaving things free to continue in the same order as they would have before the reset.  It didn't strike me as important at the time of writing that we strictly ensure all nodes get exactly the same traffic (literally 1 in a million delta :D)....

If you still don't see the value in the proposed commit, and you don't mind writing up the code you have in mind and shipping over the commit, I'm sure all us Jest users would benefit from a quick fix to this one.  Thanks!

**Comment by alkiskal** [comment]
@joeslice No please leave your PR as is. Let the admins decide which implementation fit their needs better and just reject the other.

**Comment by kramer** [comment]
I appreciate the active discussion on this issue and didn't want to disturb it as I didn't really had the time to look into the details. I liked @jasonculverhouse's suggestion with the `ImmutableList.copyOf(servers)`, as it seems to be the least complicated solution, and I'd be happy to accept a PR with that otherwise I'll apply the suggested changes on weekend.


Thank you @alkiskal for bringing up the issue and thank you @joeslice and @jasonculverhouse for chipping in!

**Comment by joeslice** [comment]
Okay thanks @kramer I will work up a fresh PR with the consolidated changes (+ test) for each of 1.0.x and master.

**Comment by joeslice** [comment]
Thanks @alkiskal @jasonculverhouse and @kramer.  I have opened #316 (and backport #317) based on the above suggestions.

**Comment by alkiskal** [comment]
Since we are thanking everyone :) kudos to @peterhagelund for pointing the guava implementation as the culprit of this issue. 

I'm closing my pull request and letting @joeslice take over. We will be depending on this library heavily in production very soon, so it would be great if this was patched up soon. @kramer Thank you for this amazing library once again. It made my life a whole lot easier and I'll make sure to contribute more in the future. Have fun all.
